src/Data/Text/Utf8/Unlower.hs

```diff
@@ -0,0 +1,2864 @@
+
+module Data.Text.Utf8.Unlower
+  (
+    unlowerCodePoint
+
+  , printUnlowerings
+  , refUnlowerCodePoint
+  ) where
+
+import Control.Monad (forM_)
+
+import qualified Data.Char as Char
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.List as List
+
+
+-- | Inverse of Char.toLower/Utf8.lowerCodePoint
+--
+-- Returns all the characters that have the given character as their lower case, for example:
+--
+--    unlowerCodePoint 'a' == "aA"
+--    unlowerCodePoint 'A' == ""
+--    unlowerCodePoint '1' == "1"
+--    unlowerCodePoint 'i' == "İiI"
+--    unlowerCodePoint 'ß' == "ẞß"
+--
+unlowerCodePoint :: Char -> [Char]
+unlowerCodePoint c =
+  maybe [c] id $ HashMap.lookup c unlowerings
+
+
+-- | Reference implementation of unlowerCodePoint, simple but slow.
+-- This is useful fo
+refUnlowerCodePoint :: Char -> [Char]
+refUnlowerCodePoint =
+  \c -> maybe [] id $ HashMap.lookup c completeMapping
+  where
+    addUnlowering c hm =
+      HashMap.insertWith (++) (Char.toLower c) [c] hm
+    completeMapping =
+      List.foldl' (flip addUnlowering) HashMap.empty [minBound..maxBound]
+
+
+-- | This function generates the definition of 'unlowerings', so that you can paste it below.
+printUnlowerings :: IO ()
+printUnlowerings = do
+
+  let
+    showCP :: Char -> String
+    showCP c = case Char.ord c of
+      co | co > 68000 -> show co -- Some RTL languages above these code points are annoying to print
+      co -> c : " (" <> show co <> ")"
+
+    showCPs :: [Char] -> String
+    showCPs cs = List.intercalate " " (map showCP cs)
+
+    charTuple lc = (lc, refUnlowerCodePoint lc)
+    includeInHashMap (lc, ucs) = ucs /= [lc]   -- Include in hashmap if it's not just id
+
+    category (lc, ucs)
+      | ucs == [lc, Char.toUpper lc] = Simple
+      | ucs == [Char.toUpper lc, lc] = Simple
+      | ucs == [] = AlreadyUppercase
+      | otherwise = Special
+
+    lst :: [(Char, [Char])]
+    lst = List.sortOn ((,) <$> category <*> fst) $ filter includeInHashMap $ map charTuple $ [minBound..maxBound]
+
+  putStrLn "  HashMap.fromList"
+  forM_ (zip ('[':repeat ',') lst) $ \(prefixChar, (lc, ucs)) -> do
+    let
+      comment = case category (lc, ucs) of
+        Special -> "SPECIAL: " <> showCP lc <> " -> " <> showCPs ucs
+        AlreadyUppercase -> "ALREADY UPPERCASE: " <> showCP lc
+        Simple -> showCP lc <> " -> " <> List.intercalate " " (map showCP ucs)
+    -- putStrLn $ "    " <> [prefixChar] <> " (" <> show lc <> ", " <> show ucs <> "\")  -- " <> comment
+    putStrLn $ "    " <> [prefixChar] <> " " <> show (lc, ucs) <> "  -- " <> comment
+  putStrLn "    ]"
+
+data UnloweringCategory
+  = Special
+  | AlreadyUppercase
+  | Simple
+  deriving (Show, Ord, Eq)
+
+
+
+-- | This hardcoded data is exhaustively tested (because we can easily iterate all codepoints) by
+-- comparing 'unlowerCodePoint' with the reference implementation.
+unlowerings :: HashMap.HashMap Char [Char]
+unlowerings =
+  HashMap.fromList
```

Comment:
Yeah I'll change it to a lazy computation instead. In ghci the first call to `refUnlowerCodePoint` seems to take about a second or so, and after that it's fast.

In an earlier version I used `unlowerCodePoint` in the `minimumSkipForCodePoint` implementation, so it would get called very often. Now it's just used in the reference implementation `refMinimumSkipForCodePoint` and in the `needleCasings` function, both are not used in production.

I liked the clear overview that this explicit list gives, but the implementation of `Char.toLower` differs between `base` versions so that was giving trouble in the shm build (https://channable.semaphoreci.com/jobs/8b16f59a-eb8c-4302-868d-9de27e517a3a). This will likely change more often so it'll be a maintenance burden.